### PR TITLE
[3.3.5][6.x]Scripts/Karazhan: Implement Optional Boss Spawn

### LINKED ISSUE #15756: [3.3.5][6.x][Karazahn] Missing Optional Boss - Hyakiss the Lurker
rev: https://github.com/TrinityCore/TrinityCore/commit/b2575a04ce076aaaa0d0cb38737950b202127132

http://www.wowhead.com/npc=16179/hyakiss-the-lurker#comments:id=37735

By huntergenwedar (2,138 – 4·15) on 2007/02/18 (Patch 2.0.8)		
This boss is one of three random bosses that can spawn in the basement (Servants Quarter) of Karazhan once you have killed all of the spiders, bats and hounds in the area. You must kill ALL of them and have the entire area clear, no respawns for one to spawn.

The other bosses are:

http://www.wowhead.com/?npc=16181#z0z (hound, Rokad the Ravager)
http://www.wowhead.com/?npc=16179#z0z (bat, Shadikith the Glider)

---
Some datatypes are already set in the instancescript, but there is NO script at all for the summon.
https://github.com/TrinityCore/TrinityCore/blob/3bc3b67a8fae059e1019446b09882dfa7fcd11a0/src/server/scripts/EasternKingdoms/Karazhan/instance_karazhan.cpp#L121

I think the stuff must be handled like the script for 'the wild eck' in gundrak. if all pre-adds are defeated a boss summons (aka if there is no guid left, npc xy spawns). Here we have the additional difficulty that there is a possiblity that there will be no bossspawn at all. This must be implemeted too.

### PR DESCRIPTION
**Pull request by Lumber-Jack-2**
- closes #15756 

http://www.wowhead.com/npc=16179/hyakiss-the-lurker#comments:id=37735

By huntergenwedar (2,138 – 4·15) on 2007/02/18 (Patch 2.0.8)  
This boss is one of three random bosses that can spawn in the basement (Servants Quarter) of Karazhan once you have killed all of the spiders, bats and hounds in the area. You must kill ALL of them and have the entire area clear, no respawns for one to spawn.

The other bosses are:

http://www.wowhead.com/?npc=16181#z0z (hound, Rokad the Ravager)
http://www.wowhead.com/?npc=16179#z0z (bat, Shadikith the Glider)

### CODE REVIEW

**Comment by Rushor** [comment]
pls don't force me to update the whole instancescript, i know it very ugly how it is currently :D

File: src/server/scripts/EasternKingdoms/Karazhan/instance_karazhan.cpp [hunk 1]
```diff
@@ -118,7 +150,28 @@ class instance_karazhan : public InstanceMapScript
                     m_auiEncounter[1] = uiData;
                     break;
                 case TYPE_MAIDEN:               m_auiEncounter[2] = uiData; break;
-                case TYPE_OPTIONAL_BOSS:        m_auiEncounter[3] = uiData; break;
+                case TYPE_OPTIONAL_BOSS:
+                    m_auiEncounter[3] = uiData;
+                    if (uiData == NOT_STARTED)
+                    {
+                        ++OptionalBossCount;
+                        if (OptionalBossCount == 50)
```
Comment (DDuarte) [inline src/server/scripts/EasternKingdoms/Karazhan/instance_karazhan.cpp]: Why 50?

Comment (Rushor) [inline src/server/scripts/EasternKingdoms/Karazhan/instance_karazhan.cpp]: @DDuarte there are 50 trashmobs, if all of them are dows one randomboss must spawn

**Comment by Keader** [comment]
>/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/World/npcs_special.cpp:2669:35: fatal error: 
      unused parameter 'player' [-Wunused-parameter]
        void ReceiveEmote(Player* player, uint32 emote) override

Why travis is complaining of a file that is not part of this PR?

**Comment by Aokromes** [comment]
Because he pushed the fix before vm fixed the warning.

**Comment by Rushor** [comment]
travis fixed by someone

the script itself works fine - going to merge it in 2 days if noone has anything to add here.

**Pull request merged**